cpp/include/raft/neighbors/cagra.cuh

```diff
@@ -404,6 +405,87 @@ void search_with_filtering(raft::resources const& res,
     res, params, idx, queries_internal, neighbors_internal, distances_internal, sample_filter);
 }
 
+/**
+ * @brief Search ANN using the constructed index with the given sample filter.
+ *
+ * Usage example:
+ * @code{.cpp}
+ *   using namespace raft::neighbors;
+ *   // memory space for a new dataset and graph
+ *   auto updated_dataset = raft::make_host_matrix<uint32_t,int64_t>(res,updated_dataset_size,dim);
+ *   update_dataset(updated_dataset);
+ *   auto updated_graph   = raft::make_host_matrix<float,int64_t>(res, updated_dataset_size,
+ * graph_degree);
+ *
+ *   // update graph
+ *   const uint32_t update_batch_size = 100;
+ *   cagra::add_graph_nodes(res, update_dataset.view(), index, update_batch_size,
+ * updated_graph.view());
+ *   // update index
+ *   index.update_graph(updated_graph);
+ *   index.update_dataset(updated_dataset);
+ * @endcode
+ *
+ * @tparam T data element type
+ * @tparam IdxT type of the indices
+ *
+ * @param[in] handle raft resources
+ * @param[in] updated_dataset a host memory for updated dataset [new_dataset_size,
+ * original_index->dim()]
+ * @param[in] max_batch_size the batch size for processing new
+ * @param[out] updated_graph a host memory for updated graph [new_dataset_size,
+ * original_index->graph_degree()]
+ */
+template <class T, class IdxT>
+void add_graph_nodes(raft::resources const& handle,
+                     const raft::device_matrix_view<const T, std::int64_t> updated_dataset,
+                     const raft::neighbors::cagra::index<T, IdxT>& original_index,
+                     const std::size_t max_batch_size,
+                     raft::host_matrix_view<IdxT, std::int64_t> updated_graph)
+{
+  assert(updated_dataset.extent(0) >= original_index.size());
+
+  const std::size_t num_new_nodes        = updated_dataset.extent(0) - original_index.size();
+  const std::size_t initial_dataset_size = original_index.size();
+  const std::size_t degree               = original_index.graph_degree();
+  const std::size_t dim                  = original_index.dim();
+  const std::size_t max_batch_size_      = max_batch_size == 0 ? num_new_nodes : max_batch_size;
```

Comment:
No, it is used to control the balance between graph quality and computation time. In this implementation, a new vector is not connected to vectors in the same batch. Therefore, decreasing the batch size improves connectivity. However, we can get higher throughput by increasing the batch size because we can run the CAGRA search with a large batch size. I think the decrease in performance and the improvement in graph quality depends on the dataset.